Announcing Spin v4.0 (#198)
spinframework/spin-docs: content/blog/announcing-spin-4-0.md

title = "Announcing Spin v4.0"
date = "2026-04-23T17:00:00Z"
template = "blog_post"
description = "Announcing Spin v4.0: stabilized WASIp3 support, async host APIs across the board, build profiles, and fine-grained capability inheritance for dependencies."
tags = []

[extra]
type = "post"
author = "The Spin Project"

---

> The CNCF Spin project just released [Spin v4.0.0](https://github.com/spinframework/spin/releases/tag/v4.0.0), the first major release of Spin in over a year, and the release we've been building toward all through the 3.x series. This release turns WASIp3 from an experimental, opt-in toggle into a first-class, long-term-supported feature; rewrites Spin's host APIs around `async`; adds build profiles; and introduces fine-grained capability inheritance for component dependencies.

Reviewer: ```suggestion
The CNCF Spin project just released [Spin v4.0.0](https://github.com/spinframework/spin/releases/tag/v4.0.0), the first major release of Spin in over a year. This release delivers a production-grade, stable implementation of WASI Preview 3; rewrites Spin's host APIs around `async`; adds build profiles; and introduces fine-grained capability inheritance for component dependencies.
```

If you want to keep something about "building towards," it's really only over the last two releases of Spin I think?  But I am wary of presenting a year's worth of Spin 3 goodies as stepping stones on the way to Spin 4 (even though I get how it can feel like that right now!).

There's a lot in this release, so this post is part release-notes and part tutorial. We'll walk through each headline feature with a working example.

> - [WASIp3: stabilized and supported long-term](#wasip3-stabilized-and-supported-long-term)

Reviewer: ```suggestion
- [WASI Preview 3: stabilized and supported long-term](#wasip3-stabilized-and-supported-long-term)
```

- [Async everywhere: Spin's host interfaces are now async](#async-everywhere-spins-host-interfaces-are-now-async)
- [Build profiles](#build-profiles)
- [Fine-grained capability inheritance for dependencies](#fine-grained-capability-inheritance-for-dependencies)
- [Upgrading to Spin 4.0](#upgrading-to-spin-40)

> ## WASIp3: stabilized and supported long-term

Reviewer: ```suggestion
## WASI Preview 3: stabilized and supported long-term
```

> WASIp3 is the next major revision of the WebAssembly System Interface. It brings first-class async, concurrent component exports, and significantly simpler WIT definitions to the component model. If you followed along in [Spin 3.5](https://spinframework.dev/v3/blog/announcing-spin-3-5) and [Spin 3.6](https://spinframework.dev/v3/blog/announcing-spin-3-6), you've seen WASIp3 progress from "experimental, opt-in, might break between releases" to something much closer to production.

Reviewer: ```suggestion
WASI Preview 3 (WASIp3) is the next major revision of the WebAssembly System Interface. It brings first-class async, concurrent component exports, and significantly simpler WIT definitions to the component model. If you followed along in [Spin 3.5](https://spinframework.dev/v3/blog/announcing-spin-3-5) and [Spin 3.6](https://spinframework.dev/v3/blog/announcing-spin-3-6), you've seen WASIp3 progress from "experimental, opt-in, might break between releases" to something much closer to production.
```

> WASIp3 is the next major revision of the WebAssembly System Interface. It brings first-class async, concurrent component exports, and significantly simpler WIT definitions to the component model. If you followed along in [Spin 3.5](https://spinframework.dev/v3/blog/announcing-spin-3-5) and [Spin 3.6](https://spinframework.dev/v3/blog/announcing-spin-3-6), you've seen WASIp3 progress from "experimental, opt-in, might break between releases" to something much closer to production.

Reviewer: I am not sure about leading on "if you followed 3.5 and 3.6."  A major release is more likely to get the attention of readers who _didn't_ follow the minutiae of minor versions.  How does this read if we keep the first para as "we have WASIp3 and this is why you should care" and relegate the history to a later paragraph?

> **Spin 4.0 ships with the March 2026 release candidate of WASIp3, and we are committing to supporting it long-term.** You no longer have to flip an experimental executor switch, WASIp3 is now the default path for new applications, and `spin-sdk`, the Spin Python SDK, and the Spin Go SDK have all been updated to use it.

Reviewer: Again, I would skip the Bad Things You No Longer Need To Do.

```suggestion
**Spin 4.0 ships with the March 2026 release candidate of WASIp3, and we are committing to supporting it long-term.** WASIp3 is now the default platform for new applications, and the Spin Rust, Python, and Go SDKs have all been updated to use it.
```

What this means in practice:

> - **No more `executor = { type = "wasip3-unstable" }`.** The HTTP trigger speaks WASIp3 natively. WASIp2 components continue to run unchanged.

Reviewer: Move this to the last bullet point and qualify it (so readers aren't going "wait what what executor I never heard about that") e.g.

```suggestion
- If you tried the WASIp3 betas in Spin 3.5 and 3.6, you'll be glad to hear this one: **no more unstable executor declarations, no more SDK feature flags** The HTTP trigger speaks WASIp3 natively. WASIp2 components continue to run unchanged.
```

actually that last sentence merits being its own bullet point

- **Concurrent, async component exports.** A single component instance can service multiple in-flight requests concurrently, instead of one instance per request.
- **One idiomatic story per language.** Rust handlers are `async fn` using `http` / `hyper` types. Python handlers are `async def`. Go handlers use standard `net/http`.

Here's the new minimum-viable Rust HTTP component in 4.0:

```rust
use spin_sdk::http::{IntoResponse, Request, Response};
use spin_sdk::http_service;

#[http_service]
async fn handle(_req: Request) -> anyhow::Result<impl IntoResponse> {
    Ok(Response::builder()
        .status(200)
        .header("content-type", "text/plain")
        .body("Hello, Spin 4!".to_string())?)
}
```

A few things to notice:

- The handler is `async fn`. That's not cosmetic, it's a real WASIp3 async export. While this handler is awaiting I/O, Spin can dispatch another request into the same component instance.

> - `Request` and `Response` are re-exports from the ecosystem `http` crate, so this code composes with Axum, Tower, `http-body-util`, and friends with no glue types.

Reviewer: The `grpc` sample is a really powerful example of this - maybe worth linking to it to illustrate?

> - `Request` and `Response` are re-exports from the ecosystem `http` crate, so this code composes with Axum, Tower, `http-body-util`, and friends with no glue types.

Reviewer: It's not visible in this code, because Spin SDK has always allowed you to hide it, but: no `ResponseOutparam` nonsense.  It's request-response.

(Although this is language specific I guess.  Go still surfaces it via `ResponseWriter` because that is the Right Thing in Go.)

Author: That's true but i'm not sure worth mentioning since Spin devs weren't super exposed to those types.

> - There's no `wasip3-unstable` feature flag in `Cargo.toml` anymore.

Reviewer: (apologies for harping on this)

For most readers, there _never was_ a `wasip3-unstable` feature flag.  I think we should be focusing our comparisons on P2, not on P3-double-preview.

### Python and Go

The same story holds in Python and Go. The Python SDK exposes a single `handle_request` coroutine:

```python
from spin_sdk.http import Handler, Request, Response

class HttpHandler(Handler):
    async def handle_request(self, request: Request) -> Response:
        return Response(
            200,
            {"content-type": "text/plain"},
            bytes("Hello from Python on WASIp3!", "utf-8"),
        )
```

Build it with:

```bash
componentize-py -w spin:up/http-trigger@4.0.0 componentize app -o app.wasm
```

> On the Go side, the 4.0 release lines up with Go SDK `v3`, which drops the TinyGo requirement. Spin Go components now build with the standard Go toolchain (Go 1.25.5+) via `componentize-go`:

Reviewer: "drops the TinyGo requirement" feels like a significant enough thing that it should be highlighted a bit more than the middle of a paragraph, but I am not sure how to organise this information (since the rest of the Go SDK stuff is pretty "it just works by magic")

```go
package main

import (
    "fmt"
    "net/http"

    spinhttp "github.com/spinframework/spin-go-sdk/v3/http"
)

func init() {
    spinhttp.Handle(func(w http.ResponseWriter, r *http.Request) {
        w.Header().Set("content-type", "text/plain")
        fmt.Fprintln(w, "Hello from Go on WASIp3!")
    })
}

func main() {}
```

```toml
[component.hello-go.build]
command = "go tool componentize-go build"
```

> If you've been writing Spin Go components against `spin-go-sdk/v2` with TinyGo, this is the big one, standard Go, standard `net/http`, and no `tinygo build -target=wasip1 -gc=leaking ...` incantation.

Reviewer: Go has always been standard `net/http`.

        Either::Right((len, _)) => ("Component model book", len?),
    };

    Ok(format!("{winner} responded first, content-length = {len:?}\n"))
}

async fn content_length(url: &str) -> anyhow::Result<Option<u64>> {
    let req = Request::get(url).body(EmptyBody::new())?;
    let res = send(req).await?;
    Ok(res
        .headers()
        .get("content-length")
        .and_then(|v| v.to_str().ok())
        .and_then(|v| v.parse().ok()))
}
```

Both requests are in flight at once, inside a single Wasm instance, scheduled by Spin. That same instance may also be serving other requests concurrently.

## Async everywhere: Spin's host interfaces are now async

WASIp3 unlocks async, but the benefit only lands if the *host APIs* your component calls are also async. A lot of Spin 4.0's work happened here: we've asyncified Spin's host interfaces so I/O-heavy handlers actually get concurrency instead of blocking the instance.

In 4.0, the following interfaces are async from the guest's perspective:

- **Key-value**: `Store::open_default().await`, `store.get(key).await`, `store.set(key, value).await`
- **SQLite**: `Connection::open_default().await`, `connection.execute(...).await`

> - **PostgreSQL**: new async client API

Reviewer: It's a bit weird that the others have examples and this doesn't?  I don't think anything particularly difficult or complicated is going on that the level shown here, I think it's ye olde `Connection::open(...).await`, `connection.query(...).await`, no?

Author: Oh good catch, i meant to come back to this one!

- **Redis (outbound and trigger)**: `Connection::open(...).await`, plus async Redis subscriber handlers
- **Outbound HTTP**: `spin_sdk::http::send(req).await`

> Here's a key-value handler in Rust that both *uses* async host APIs and *produces* a streaming HTTP response via `spin_sdk::wasip3::spawn`:

Reviewer: I felt the lack of streaming responses in the HTTP section - perhaps we could foreshadow this sample there?

Author: Foreshadowed!

```rust
use bytes::Bytes;
use futures::{SinkExt, StreamExt};
use http_body_util::StreamBody;
use spin_sdk::http::{IntoResponse, Request, Response};
use spin_sdk::http_service;

#[http_service]
async fn handle(_req: Request) -> anyhow::Result<impl IntoResponse> {
    let store = spin_sdk::key_value::Store::open_default().await?;

    // A channel the spawned task will push body chunks into.
    let (mut tx, rx) = futures::channel::mpsc::channel::<Bytes>(1024);
    let rx = rx.map(|value| anyhow::Ok(http_body::Frame::data(value)));
    let response = Response::new(StreamBody::new(rx));

    // Spawn a background Wasm task. It outlives `handle` returning.
    spin_sdk::wasip3::spawn(async move {

>         if let Ok(Some(greeting)) = store.get("greeting").await {

Reviewer: This example is clear and minimal as a "how to", but it's also not particularly compelling as a "why you should care."  I wonder if a SQLite example that does a query and streams rows might make for a better story?  Probably a nit though.

            let _ = tx.send(greeting.into()).await;
        }
        if let Ok(Some(who)) = store.get("greetee").await {
            let _ = tx.send(" ".into()).await;
            let _ = tx.send(who.into()).await;
        }
        // Dropping `tx` closes the body stream.
    });

    Ok(response)
}
```

> Spin starts streaming the response as soon as `handle` returns, while the spawned task keeps reading from the key-value store in the background. In the WASIp2 world, each of those KV reads would have blocked the entire component instance. In 4.0, they don't.

Reviewer: We are `await`-ing the reads, so they are kind of blocking execution anyway.  The highlight here (for me) is more that we are returning a response _immediately_ and then _writing to the response after it has begun._  It's tying in with the whole practical thing of time to first byte or whatever the Web folks call it - not needing to wait for _all_ the reads to complete before sending the ones we have.

Of course that doesn't really tie in with the "behold, we are showing you async host interfaces" story that this section is about...

The same pattern is available in Python (via `componentize_py_async_support.spawn`) and Go (via plain `go func() { ... }()` goroutines).

> > **Heads up on global state.** Because a single instance can now serve multiple concurrent requests, instance-scoped "global" state is shared across in-flight requests. This is the same rule you already follow in Axum, Express, or Flask, but it's a genuine shift from "one instance per request" if you've been writing Spin components against WASIp2 semantics. Audit any `static`, module-level, or `OnceCell` state and reach for per-request state or explicit synchronization where needed.

Reviewer: I like this (more concise and yet more clear than my yammering) but I am not sure that the "Spin host interfaces are async" section is the right place for it.

Author: Yeah, I wasn't quite sure where to put this but relative to everything else in this post, the "streaming" context seemed most appropriate. Let me know if you have a better suggestion! I could move it up to the end of the concurrent outbound http section above?

Reviewer: Maybe cop out and make "A heads up on global state" a new subheading in the P3 section?

Author: I moved this to a new subheading at the end of `WASI Preview3: ...` section just after `Concurrent outbound HTTP: ...`. Reason: We extoll the concurrency implications at the end of this section so to me it flows a bit better but happy to shuffle this around if you think it should be more prominently featured further up.

## Build profiles

> Until now, building a Spin component in debug vs. release mode, or profiling builds, or "use a pre-built version in CI", meant hand-editing `[component.X.build]` tables, usually on a branch you hoped nobody merged. [SIP-022](https://github.com/spinframework/spin/blob/main/docs/content/sips/022-build-profiles.md) fixes this. Spin 4.0 introduces named **build profiles**, inspired by Cargo profiles.

Reviewer: ```suggestion
Development tools often offer ways to build code in different ways for different purposes: for example, debug builds that favour diagnostic value over speed and size, or profiling builds that inject performance instrumentation.  Spin's build profiles provide a way for you to leverage those options in Wasm applications - for example, to define a debug or profile build of your application.
```

(Reasoning: SIPs are hardly airport thrillers material, Go programmers don't know what Cargo is, and "look how bad you've had it for the last five years" has made me makes me flinch ever since someone pointed out how much Microsoft used this formula (usually when announcing another soon-to-be-deprecated data stack).)

```toml
spin_manifest_version = 2

[application]
name = "sentiment-analysis"

[component.sentiment-analysis]
source = "target/spin-http-js.wasm"

[component.sentiment-analysis.build]
command = "npm run build"
watch = ["src/**/*", "package.json", "package-lock.json"]

# A `debug` profile for the sentiment-analysis component.
[component.sentiment-analysis.profile.debug]
source = "target/spin-http-js.debug.wasm"

[component.sentiment-analysis.profile.debug.build]
command = "npm run build:debug"

# The `ui` component has no debug profile, it'll fall back to the default.
[component.ui]
source = { url = "https://.../spin_static_fs.wasm", digest = "sha256:..." }

# The `kv-explorer` component pulls a pre-built debug build from a registry.
[component.kv-explorer]
source = { url = "https://.../spin-kv-explorer.wasm", digest = "sha256:..." }

[component.kv-explorer.profile.debug]
source = { url = "https://.../spin-kv-explorer.debug.wasm", digest = "sha256:..." }
```

Now you can run the same manifest in two modes:

```bash
# Production: every component uses its default build.
$ spin up

# Debug: every component that defines `debug` uses it; others fall back.
$ spin up --profile debug
```

> The `--profile` flag is recognized by `spin build`, `spin watch`, `spin up`, `spin deploy`, and `spin registry push`, so profiles carry all the way from local dev through deployment.

Reviewer: We can't really claim it for `spin deploy` because that depends on the deployment plugin.  I am not sure where that leaves the last clause of the sentence - I see what you're getting at, that you can put e.g. a profiling build into your production environment, but I am not sure how to express it when this is really the purview of the deployment tool.

Author: Yeah good catch. I need to purge `spin deploy` as it existed from my mental model 😆 Should i just purge this paragraph and re-direct to the docs for more info? Or i can say something more generic in the last clause like: "so profiles carry from local development into your published artifacts"

Reviewer: I think you're right, let's purge it rather than trying to craft something delicate.  This is an announcement: it's okay not to dive too deep.

> The fields a profile can override are intentionally scoped to build-time concerns:

Reviewer: To me, this and the "not atomic" para are more detail than you need in an announcement. Maybe a "Learn more in the Spin docs" link instead?

- `source`
- `build.command`
- `environment`
- `dependencies`

Profiles are *not* atomic, fields in a profile are individual overrides. If a component doesn't define a requested profile, it falls back to its default configuration rather than erroring. This is exactly what you want for mixed-language apps, where only some components have meaningful debug builds.

## Fine-grained capability inheritance for dependencies

> [SIP-020](https://github.com/spinframework/spin/blob/main/docs/content/sips/020-component-dependencies.md) introduced component dependencies, along with a single component-level boolean: `dependencies_inherit_configuration`. It was all-or-nothing, either every dependency inherited every capability of the parent component, or none did.

Reviewer: ```suggestion
Spin 3 introduced component dependencies, along with a single component-level boolean: `dependencies_inherit_configuration`. It was all-or-nothing, either every dependency inherited every capability of the parent component, or none did.
```

(or Spin 2.x or whatever - but a SIP number is "I can sense that's some ghastly spec without even clicking the link" levels of discouraging)

> [SIP-020](https://github.com/spinframework/spin/blob/main/docs/content/sips/020-component-dependencies.md) introduced component dependencies, along with a single component-level boolean: `dependencies_inherit_configuration`. It was all-or-nothing, either every dependency inherited every capability of the parent component, or none did.

Reviewer: Again the "it was all or nothing, that's a problem" style is a bit "look how bad we've made you have it".  How about e.g.

```suggestion
Spin 3 introduced component dependencies, which enabled you to use off-the-shelf Wasm components as libraries without having to write composition scripts or commands. Once you start using component dependencies in this way, a natural concern is what capabilities those dependencies have - your application needs to connect to your database, but your regex matcher sure doesn't.

Spin 3 offered a partial solution where dependencies could be fully isolated, and Spin 4 improves on this by allowing you to manage dependency capabilities more selectively - you can choose on a dependency-by-dependency, capability-by-capability basis what each dependency can inherit.
```

then go into your motivating example and case illustrations

> That's a problem in the real world. You might be happy letting an `aws:client/s3` dependency make outbound HTTPS calls to S3, but not letting it read your key-value store or invoke an LLM. [SIP-023](https://github.com/spinframework/spin/blob/main/docs/content/sips/023-fine-grained-capability-inheritance.md) replaces that coarse toggle with a per-dependency `inherit_configuration` field that accepts three forms.

Reviewer: I wonder if we should shy away from seemingly-real examples like `aws:client` (unless a `aws:client` package really exists)?  The "I need to make calls to S3" is nicely concrete though.

> **1. Inherit everything** (equivalent to the old global flag, but scoped to one dep):

Reviewer: ```suggestion
**1. Inherit everything** - the dependency has access to all the capabilities of the main component:

(syntax)

This is similar to `dependencies_inherit_configuration = true` in Spin 3, but scoped to just this dependency.
```

```toml
[component."infra-dashboard".dependencies]
"aws:client" = { version = "1.0.0", inherit_configuration = true }
```

**2. Inherit nothing** (the default):

```toml
[component."infra-dashboard".dependencies]
"aws:client" = { version = "1.0.0", inherit_configuration = false }
```

> The dependency is fully isolated from the parent's configuration. All capability imports are satisfied by deny adapters.

Reviewer: "All capability imports are satisfied by deny adapters" is very implementation detail.  Drop it or replace it with something user-visible ("any import the component calls returns an error").

> **3. Inherit a specific subset.** This is the new power:

Reviewer: Component-by-component isolation is also new?

```toml
[component."infra-dashboard"]
allowed_outbound_hosts = ["https://s3.us-west-2.amazonaws.com"]
key_value_stores      = ["my-key-value-cache"]

> ai_models             = ["llama2-chat"]

Reviewer: I would maybe not draw too much attention to `ai_models` since Spin 4 support for it has not really caught up to where we would like it to be...!

[component."infra-dashboard".dependencies]
"aws:client" = { version = "1.0.0", inherit_configuration = ["allowed_outbound_hosts"] }
```

> Here `aws:client` can make outbound HTTPS calls to the parent's allowed hosts, enough to reach S3, but it *cannot* see `my-key-value-cache` and *cannot* invoke `llama2-chat`. Every other capability is denied.

Reviewer: The phrasing of those last two clauses reads to me like "other" is "except AOH, KV and LLM" which makes me go "wait what is the status of KV and LLM then, are they not denied."  How about something like:

```suggestion
Here `aws:client` can make outbound HTTPS calls to the parent's allowed hosts, enough to reach S3.  But every other capability is denied. Specifically, the AWS client library *cannot* see `my-key-value-cache` and *cannot* invoke `llama2-chat`. 
```

> The keys you can list are the configuration families Spin already knows about: `ai_models`, `allowed_outbound_hosts`, `environment`, `files`, `key_value_stores`, `sqlite_databases`, and `variables`. Each key maps to a set of WIT interfaces, for example, listing `allowed_outbound_hosts` covers `wasi:http`, `fermyon:spin/mysql`, `fermyon:spin/postgres`, `fermyon:spin/redis`, `wasi:sockets`, and so on. See the [SIP-023 table](https://github.com/spinframework/spin/blob/main/docs/content/sips/023-fine-grained-capability-inheritance.md) for the full mapping.

Reviewer: This is a _lot_ of detail.

Also, don't point to the SIP.  Point to the docs, which are written more from a user perspective (whereas SIPs are design/implementation perspective), and have a much better chance of being maintained!  (And if the information isn't in the docs, then let's add it to the docs - please raise an issue if I overlooked this.)

Author: https://github.com/spinframework/spin-docs/issues/199

Stubbed this out with a TODO link for now. I'll take a pass at updating the composition docs tomorrow!

Author: Replaced the TODO with a link to `https://spinframework.dev/v4/writing-apps#using-component-dependencies`

> `inherit_configuration` works uniformly across every dependency source type:

Reviewer: Again, I fear this starts to drown the key point in detail.  This is an announcement post, not a comprehensive reference to dependency syntax.  If we want to do a post with this level of detail, let's make it a separate follow-up.

```toml
[component."my-app".dependencies]
# Registry dependency
"aws:client"      = { version = "1.0.0", inherit_configuration = ["allowed_outbound_hosts"] }
# Local path dependency
"my:lib/utils"    = { path = "lib/utils.wasm", inherit_configuration = true }
# HTTP dependency
"vendor:dep/api"  = { url = "https://example.com/dep.wasm", digest = "sha256:abc123", inherit_configuration = ["variables"] }
# Reference to another component in this same app
"infra:dep/svc"   = { component = "svc-component", inherit_configuration = ["key_value_stores", "variables"] }
```

> A couple of rules to be aware of:

Reviewer: Neither of these seems important to highlight at this level (although I'm open to persuasion on "DIC is still supported").

- The shorthand form (`"fizz:buzz" = ">=0.1.0"`) doesn't support `inherit_configuration`, use the full table form.
- `dependencies_inherit_configuration = true` still works as a convenience for "turn it on for every dependency," but **mixing** it with per-dependency `inherit_configuration` is a hard error. Pick one.

> Practically, this means you can hand out dependencies to components from across your organization, or from the registry, and grant them exactly the capabilities they need, and nothing more. It's the principle of least privilege, expressed in `spin.toml`.

Reviewer: YES THIS

Maybe put this further up, in the "explaining why" part?  I appreciate you want a nice wrap-up but many readers will have skipped to the next section by now.

Author: Good tip! I refactored this whole section to highlight this.

```toml
[component."infra-dashboard".dependencies]
"acme:s3-client" = { version = "1.0.0", inherit_configuration = true }
```

This is similar to `dependencies_inherit_configuration = true` in Spin 3, but scoped to just this dependency.

**2. Inherit nothing** (the default):

```toml
[component."infra-dashboard".dependencies]
"acme:s3-client" = { version = "1.0.0", inherit_configuration = false }
```

The dependency is fully isolated from the parent's configuration, any capability import the component calls will return an error. Per-dependency isolation like this is itself new in Spin 4.

**3. Inherit a specific subset.** This is the new power:

```toml
[component."infra-dashboard"]
allowed_outbound_hosts = ["https://s3.us-west-2.amazonaws.com"]
key_value_stores      = ["my-key-value-cache"]

[component."infra-dashboard".dependencies]
"acme:s3-client" = { version = "1.0.0", inherit_configuration = ["allowed_outbound_hosts"] }
```

Here `acme:s3-client` can make outbound HTTPS calls to the parent's allowed hosts, enough to reach S3. But every other capability is denied. Specifically, the S3 client *cannot* see `my-key-value-cache`.

> For the full list of configuration keys you can inherit and how they map to WIT interfaces, see the [Spin docs on component dependencies](TODO).

Author: TODO: https://github.com/spinframework/spin-docs/issues/199
